File: Queue Using Two Stacks/queueUsingTwoStacks.cpp
#include <bits/stdc++.h>
using namespace std;

class Queue {
    
	stack<int> s1, s2;
    
    public:
	// Enqueue an item to the queue
	void enQueue(int x){
	    
		// Push into stack S1
		this->s1.push(x);
	}

	// Dequeue an item from the queue
	int deQueue(){
		

		// if s2 is empty, then pop all elements of s1 and
		// push them to s2 in the order they are popped
		
		if (s2.empty()) {
			while (!s1.empty()) {
				this->s2.push(s1.top());
				this->s1.pop();
			}
		}

		// Top of S2 is the required front of the queue
		
		int X = this->s2.top();
		this->s2.pop();
		return X;
	}
};


int main(){
    
	Queue q;
	
	int queries;
	cin>>queries;
	
	while(queries--){
	    int type;
	    cin>>type;
	    
	    if(type == 1){
	        int X;
	        cin>>X;
	        q.enQueue(X);
	    }
	    
	    else{
	        cout<<q.deQueue()<<" ";
	    }
	}
	
	return 0;
}

File: Queue Using Two Stacks/queueUsingTwoStacks.java
import java.lang.*;
import java.io.*;
import java.util.*;

class Main{
//This class implements the queue data structure using two stacks.
static class Queue{
	 Stack<Integer> s1 = new Stack<Integer>();
	 Stack<Integer> s2 = new Stack<Integer>();

	private void enqueue(int x){
		while (!s1.isEmpty()){
			s2.push(s1.pop());
			//s1.pop();
		}
		s1.push(x);
		while (!s2.isEmpty()){
			s1.push(s2.pop());
			//s2.pop();
		}
	}
	private int dequeue(){
		if (s1.isEmpty()){
			System.out.println("Queue is Empty");
			System.exit(0);
		}
		int x = s1.peek();
		s1.pop();
		return x;
	}
}
//Main Function
public static void main(String[] args){
	int n = 5;
    int[] arr = {1,2,1,3,2,1,4,2};
    Queue queue = new Queue();
    for (int i=0;i<arr.length;i++){
        if (arr[i] == 1){
            queue.enqueue(arr[i+1]);
            i++;
        }else{
            System.out.println(queue.dequeue());
        }
    }
  }
}

